File: leet-code/743. Network Delay Time/NetworkDelayTime743.java
import java.util.*;

public class NetworkDelayTime743 {
    public int networkDelayTime(int[][] times, int n, int k) {
        Map<Integer, List<Integer[]>> map = new HashMap<>();
        for (int[] time : times) {
            map.putIfAbsent(time[0], new ArrayList<>());
            map.get(time[0]).add(new Integer[]{time[1], time[2]});
        }
        int[] distance = dijkstra(map, n, k);
        return isNotConnected(distance) ? -1 : getMaxValue(distance);
    }

    private int[] dijkstra(Map<Integer, List<Integer[]>> map, int maxNum, int start) {
        int[] distance = new int[maxNum + 1];
        boolean[] visit = new boolean[maxNum + 1];
        PriorityQueue<Integer[]> pq = new PriorityQueue<>(Comparator.comparingInt(o -> o[1]));

        int maxValue = Integer.MAX_VALUE;
        for (int i = 1; i <= maxNum; i++) distance[i] = maxValue;
        distance[0] = 0;
        distance[start] = 0;
        pq.add(new Integer[]{start, distance[start]});

        while (!pq.isEmpty()) {
            Integer[] startNode = pq.poll();
            if (visit[startNode[0]]) continue;
            visit[startNode[0]] = true;
            if (!map.containsKey(startNode[0])) continue;
            for (Integer[] endNode : map.get(startNode[0])) {
                if (visit[endNode[0]]) continue;
                if (distance[endNode[0]] < startNode[1] + endNode[1]) continue;
                distance[endNode[0]] = startNode[1] + endNode[1];
                pq.add(new Integer[]{endNode[0], distance[endNode[0]]});
            }
        }

        return distance;
    }

    private boolean isNotConnected(int[] distance) {
        return Arrays
                .stream(distance)
                .anyMatch(d -> d == Integer.MAX_VALUE);
    }

    private int getMaxValue(int[] distance) {
        return Arrays
                .stream(distance)
                .max()
                .getAsInt();
    }
}

File: leet-code/743. Network Delay Time/NetworkDelayTime743.kt
import java.util.*

class NetworkDelayTime743_2 {
    fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
        val map: MutableMap<Int, MutableList<IntArray>> = mutableMapOf()
        for (time in times) {
            map.putIfAbsent(time[0], mutableListOf())
            map[time[0]]!!.add(intArrayOf(time[1], time[2]))
        }

        fun dijkstra(): IntArray {
            val distance: IntArray = IntArray(n + 1)
            val visit: BooleanArray = BooleanArray(n + 1)
            var pq: PriorityQueue<IntArray> = PriorityQueue<IntArray>(compareBy { it[1] })

            for (i in distance.indices) distance[i] = Int.MAX_VALUE
            distance[0] = 0
            distance[k] = 0
            pq.add(intArrayOf(k, 0))

            while (!pq.isEmpty()) {
                val start = pq.poll()
                if (visit[start[0]]) continue
                visit[start[0]] = true
                if (!map.containsKey(start[0])) continue
                for (end in map[start[0]]!!) {
                    if (distance[end[0]] < distance[start[0]] + end[1]) continue
                    distance[end[0]] = distance[start[0]] + end[1]
                    pq.add(intArrayOf(end[0], distance[end[0]]))
                }
            }

            return distance
        }

        val distanceInfo = dijkstra()
        if (isNotConnected(distanceInfo)) return -1
        return getMaxValue(distanceInfo)
    }

    private fun isNotConnected(array: IntArray): Boolean {
        return array.any { it == Int.MAX_VALUE }
    }

    private fun getMaxValue(array: IntArray): Int {
        return array.max()
    }
}
